File: frontend/src/types/lote-comprobante.ts
export interface LoteComprobanteFila {
  id: number;
  fila_excel: number;
  comprobante_ref: string;
  estado: string;
  datos_json: Record<string, string | number | null> | null;
  mensajes_json: string[];
}

export interface LoteComprobanteGrupo {
  id: number;
  comprobante_ref: string;
  orden: number;
  estado: string;
  tipo_comprobante: number | null;
  concepto: number | null;
  punto_venta_numero: number | null;
  cliente_documento: string | null;
  cliente_razon_social: string | null;
  fecha_emision: string | null;
  fecha_servicio_desde: string | null;
  fecha_servicio_hasta: string | null;
  fecha_vto_pago: string | null;
  total_estimado: number;
  mensajes_json: string[];
  cae: string | null;
  numero_asignado: number | null;
  comprobante_id: number | null;
}

export interface LoteComprobanteGrupoDetalle extends LoteComprobanteGrupo {
  descripcion_facturada: string | null;
}

export interface LoteComprobante {
  id: number;
  nombre_archivo: string;
  archivo_hash: string;
  estado: string;
  modo_procesamiento: string;
  procesamiento_async: boolean;
  total_filas: number;
  total_grupos: number;
  grupos_validos: number;
  grupos_con_error: number;
  grupos_emitidos: number;
  grupos_fallidos: number;
  grupos_reconciliados_externos: number;
  grupos_descartados: number;
  mensaje_resumen: string | null;
  metadata_json: Record<string, unknown> | null;
  mapeo_usado_json: Record<string, unknown> | null;
  headers_detectados_json: string[] | null;
  started_at: string | null;
  finished_at: string | null;
  compactado_at: string | null;
  created_at: string;
  updated_at: string;
  empresa_id: number;
  usuario_id: number | null;
  formato_importacion_id: number | null;
  formato_importacion_version_id: number | null;
}

export interface LoteComprobanteDetalle extends LoteComprobante {
  grupos: LoteComprobanteGrupo[];
  filas: LoteComprobanteFila[];
}

export interface LoteTotalesListos {
  comprobantes: number;
  neto: number;
  iva21: number;
  iva105: number;
  total: number;
  valores_invalidos: number;
}

export interface LoteComprobanteResumen extends LoteComprobante {
  confirmacion_fecha_fiscal: string;
  mensaje_confirmacion_fecha_fiscal: string;
  confirmacion_reintento_fallidos: string;
  mensaje_confirmacion_reintento_fallidos: string;
  confirmacion_duplicado_logico: string;
  mensaje_confirmacion_duplicado_logico: string;
  cantidad_duplicados_logicos: number;
  fechas_emision_validas: string[];
  puntos_venta_validos: number[];
  totales_listos_para_emitir: LoteTotalesListos;
}

export interface LoteComprobanteGruposPage {
  items: LoteComprobanteGrupoDetalle[];
  page: number;
  per_page: number;
  total: number;
  total_pages: number;
  estado: string | null;
}

export interface LoteValidacionResponse {
  lote: LoteComprobante;
  puede_emitirse: boolean;
  requiere_background: boolean;
  mensaje: string;
}

export interface LoteProcesamientoResponse {
  lote: LoteComprobante;
  mensaje: string;
  en_progreso: boolean;
}

export interface LoteAccionResponse {
  lote: LoteComprobante;
  mensaje: string;
}

export interface ReconciliacionExternaItem {
  grupo_id: number;
  tipo_comprobante: number;
  punto_venta_numero: number;
  numero: number;
  fecha_emision: string;
  total: number;
  cae?: string;
  motivo: string;
}

export interface LoteOpcionesFechas {
  punto_venta_modo: "archivo" | "fijo";
  punto_venta_numero?: number;
  concepto_modo: "productos" | "servicios" | "archivo" | "";
  descripcion_item_modo: "archivo" | "fija" | "";
  descripcion_item_fija?: string;
  fecha_emision_modo: "archivo" | "fija" | "";
  fecha_emision_fija?: string;
  fecha_servicio_desde_modo: "archivo" | "fija" | "";
  fecha_servicio_desde_fija?: string;
  fecha_servicio_hasta_modo: "archivo" | "fija" | "";
  fecha_servicio_hasta_fija?: string;
  fecha_vto_pago_modo: "archivo" | "fija" | "";
  fecha_vto_pago_fija?: string;
}

export const ESTADOS_LOTE_NOMBRES: Record<string, string> = {
  cargado: "Cargado",
  validado: "Validado",
  en_cola: "En cola",
  con_errores: "Con errores",
  procesando: "Procesando",
  autorizado_parcial: "Autorizado parcial",
  requiere_reconciliacion: "Requiere reconciliación",
  completado: "Completado",
  cerrado_reconciliado: "Cerrado reconciliado",
  cerrado_con_descartes: "Cerrado con descartes",
  fallido: "Fallido",
};

export const ESTADOS_LOTE_COLOR: Record<string, string> = {
  cargado: "bg-gray-100 text-gray-800",
  validado: "bg-blue-100 text-blue-800",
  en_cola: "bg-sky-100 text-sky-800",
  con_errores: "bg-amber-100 text-amber-800",
  procesando: "bg-indigo-100 text-indigo-800",
  autorizado_parcial: "bg-yellow-100 text-yellow-800",
  requiere_reconciliacion: "bg-orange-100 text-orange-800",
  completado: "bg-green-100 text-green-800",
  cerrado_reconciliado: "bg-emerald-100 text-emerald-800",
  cerrado_con_descartes: "bg-stone-100 text-stone-800",
  fallido: "bg-red-100 text-red-800",
};

export const ESTADOS_GRUPO_COLOR: Record<string, string> = {
  validado: "bg-blue-50 text-blue-700",
  autorizado: "bg-green-50 text-green-700",
  autorizado_externo: "bg-emerald-50 text-emerald-700",
  con_error: "bg-amber-50 text-amber-700",
  requiere_reconciliacion: "bg-orange-50 text-orange-700",
  reintentando: "bg-orange-50 text-orange-700",
  fallido: "bg-red-50 text-red-700",
  descartado: "bg-stone-50 text-stone-700",
  cargado: "bg-gray-50 text-gray-700",
};

export const ESTADOS_GRUPO_NOMBRES: Record<string, string> = {
  validado: "Validado",
  autorizado: "Autorizado",
  autorizado_externo: "Autorizado externo",
  con_error: "Con observaciones",
  requiere_reconciliacion: "Requiere reconciliación",
  reintentando: "Requiere reconciliación",
  fallido: "Fallido",
  descartado: "Descartado",
  cargado: "Cargado",
};

File: backend/app/schemas/lote_comprobante.py
"""Schemas para emisión masiva de comprobantes."""

from datetime import date, datetime
from decimal import Decimal
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field


class LoteComprobanteFilaResponse(BaseModel):
    """Representa una fila validada del Excel."""

    id: int
    fila_excel: int
    comprobante_ref: str
    estado: str
    datos_json: dict[str, Any] | None = None
    mensajes_json: list[str] = Field(default_factory=list)

    class Config:
        from_attributes = True


class LoteComprobanteGrupoResponse(BaseModel):
    """Representa un comprobante agrupado dentro del lote."""

    id: int
    comprobante_ref: str
    orden: int
    estado: str
    tipo_comprobante: Optional[int] = None
    concepto: Optional[int] = None
    punto_venta_numero: Optional[int] = None
    cliente_documento: Optional[str] = None
    cliente_razon_social: Optional[str] = None
    fecha_emision: Optional[date] = None
    fecha_servicio_desde: Optional[date] = None
    fecha_servicio_hasta: Optional[date] = None
    fecha_vto_pago: Optional[date] = None
    total_estimado: Decimal = Decimal("0")
    mensajes_json: list[str] = Field(default_factory=list)
    cae: Optional[str] = None
    numero_asignado: Optional[int] = None
    comprobante_id: Optional[int] = None

    class Config:
        from_attributes = True


class LoteComprobanteGrupoDetalleResponse(LoteComprobanteGrupoResponse):
    """Representa un comprobante agrupado con datos derivados para la UI."""

    descripcion_facturada: Optional[str] = None


class LoteComprobanteResponse(BaseModel):
    """Representa el estado general de un lote."""

    id: int
    nombre_archivo: str
    archivo_hash: str
    estado: str
    modo_procesamiento: str
    procesamiento_async: bool
    total_filas: int
    total_grupos: int
    grupos_validos: int
    grupos_con_error: int
    grupos_emitidos: int
    grupos_fallidos: int
    grupos_reconciliados_externos: int = 0
    grupos_descartados: int = 0
    mensaje_resumen: Optional[str] = None
    metadata_json: dict[str, Any] | None = None
    mapeo_usado_json: dict[str, Any] | None = None
    headers_detectados_json: list[str] | None = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    compactado_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    empresa_id: int
    usuario_id: Optional[int] = None
    formato_importacion_id: Optional[int] = None
    formato_importacion_version_id: Optional[int] = None

    class Config:
        from_attributes = True


class LoteComprobanteDetalleResponse(LoteComprobanteResponse):
    """Incluye el detalle de grupos y filas del lote."""

    grupos: list[LoteComprobanteGrupoResponse] = Field(default_factory=list)
    filas: list[LoteComprobanteFilaResponse] = Field(default_factory=list)


class LoteTotalesListosResponse(BaseModel):
    """Totales agregados de los comprobantes válidos del lote."""

    comprobantes: int = 0
    neto: float = 0
    iva21: float = 0
    iva105: float = 0
    total: float = 0
    valores_invalidos: int = 0


class LoteComprobanteResumenResponse(LoteComprobanteResponse):
    """Resumen operativo liviano para abrir lotes grandes."""

    confirmacion_fecha_fiscal: str = ""
    mensaje_confirmacion_fecha_fiscal: str = ""
    confirmacion_reintento_fallidos: str = ""
    mensaje_confirmacion_reintento_fallidos: str = ""
    confirmacion_duplicado_logico: str = ""
    mensaje_confirmacion_duplicado_logico: str = ""
    cantidad_duplicados_logicos: int = 0
    fechas_emision_validas: list[str] = Field(default_factory=list)
    puntos_venta_validos: list[int] = Field(default_factory=list)
    totales_listos_para_emitir: LoteTotalesListosResponse = Field(
        default_factory=LoteTotalesListosResponse
    )


class LoteComprobanteGruposPageResponse(BaseModel):
    """Página de comprobantes agrupados de un lote."""

    items: list[LoteComprobanteGrupoDetalleResponse] = Field(default_factory=list)
    page: int
    per_page: int
    total: int
    total_pages: int
    estado: Optional[str] = None


class LoteValidacionResponse(BaseModel):
    """Respuesta de validación o creación del lote."""

    lote: LoteComprobanteResponse
    puede_emitirse: bool
    requiere_background: bool
    mensaje: str


class LoteProcesamientoResponse(BaseModel):
    """Respuesta al iniciar o completar la emisión de un lote."""

    lote: LoteComprobanteResponse
    mensaje: str
    en_progreso: bool


class LoteGrupoIdsRequest(BaseModel):
    """Selección explícita de grupos de un lote."""

    model_config = ConfigDict(extra="forbid")

    grupo_ids: list[int] = Field(default_factory=list)


class LoteDescartarGruposRequest(BaseModel):
    """Request para descartar grupos pendientes."""

    model_config = ConfigDict(extra="forbid")

    grupo_ids: list[int] = Field(..., min_length=1)
    motivo: str = Field(..., min_length=3, max_length=500)


class LoteEliminarCompactarRequest(BaseModel):
    """Request con motivo obligatorio para acciones destructivas o compactación."""

    model_config = ConfigDict(extra="forbid")

    motivo: str = Field(..., min_length=3, max_length=500)


class LoteReconciliacionExternaItem(BaseModel):
    """Datos confirmados de un comprobante emitido fuera de FactuFlow."""

    model_config = ConfigDict(extra="forbid")

    grupo_id: int
    tipo_comprobante: int = Field(..., ge=1, le=999)
    punto_venta_numero: int = Field(..., ge=1, le=99999)
    numero: int = Field(..., ge=1)
    fecha_emision: date
    total: Decimal = Field(..., ge=0)
    cae: Optional[str] = Field(None, min_length=1, max_length=20)
    motivo: str = Field(..., min_length=3, max_length=500)


class LoteReconciliacionExternaRequest(BaseModel):
    """Request para reconciliar comprobantes emitidos en ARCA Web."""

    model_config = ConfigDict(extra="forbid")

    comprobantes: list[LoteReconciliacionExternaItem] = Field(..., min_length=1)


class LoteAccionResponse(BaseModel):
    """Respuesta genérica de una acción resolutiva de lote."""

    lote: LoteComprobanteResponse
    mensaje: str
